gohermes/mid_hermes.go

```diff
@@ -21,55 +21,42 @@
 package gohermes
 
 /*
-#cgo LDFLAGS: -lhermes_mid_hermes -lhermes_mid_hermes_ll -lhermes_credential_store -lhermes_data_store -lhermes_key_store -lhermes_rpc -lhermes_common -lthemis -lsoter
+#cgo LDFLAGS: -lhermes_mid_hermes
 #include <hermes/mid_hermes/mid_hermes.h>
 #include <string.h>
 */
 import "C"
-
 import (
 	"errors"
-	"runtime"
+	"fmt"
 	"unsafe"
 )
 
-type MidHermes struct {
-	mid_hermes                 *C.mid_hermes_t
-	credential_store_transport *HermesTransport
-	key_store_transport        *HermesTransport
-	data_store_transport       *HermesTransport
+var ErrEmptyParameter = errors.New("empty parameter")
+
+// MidHermesWrapper hide fetching C pointer to mid_hermes from go wrappers
+type MidHermesWrapper interface {
+	GetMidHermes() unsafe.Pointer
+	Close() error
 }
 
-func MidHermes_finalize(mid_hermes *MidHermes) {
-	mid_hermes.Close()
+// MidHermes used to pass Go function call to C
+type MidHermes struct {
+	wrapper    MidHermesWrapper
+	mid_hermes *C.mid_hermes_t
 }
 
-func NewMidHermes(
-	id []byte, private_key []byte,
-	credential_store_transport *HermesTransport,
-	data_store_transport *HermesTransport,
-	key_store_transport *HermesTransport) (*MidHermes, error) {
-
-	mh := &MidHermes{
-		credential_store_transport: credential_store_transport,
-		data_store_transport:       data_store_transport,
-		key_store_transport:        key_store_transport}
-	mh.mid_hermes = C.mid_hermes_create(
-		(*C.uint8_t)(unsafe.Pointer(&id[0])), C.size_t(len(id)),
-		(*C.uint8_t)(unsafe.Pointer(&private_key[0])), C.size_t(len(private_key)),
-		mh.key_store_transport.GetHermesTransport(),
-		mh.data_store_transport.GetHermesTransport(),
-		mh.credential_store_transport.GetHermesTransport())
-	if nil == mh.mid_hermes {
-		return nil, errors.New("MidHermes object creation error")
-	}
-	runtime.SetFinalizer(mh, MidHermes_finalize)
-	return mh, nil
+// NewMidHermes return new MidHermes
+func NewMidHermes(wrapper MidHermesWrapper) (*MidHermes, error) {
+	hermes := (*C.mid_hermes_t)(wrapper.GetMidHermes())
+	return &MidHermes{wrapper: wrapper, mid_hermes: hermes}, nil
 }
 
 func (mh *MidHermes) Close() error {
-	C.mid_hermes_destroy(&(mh.mid_hermes))
-	return nil
+	// TODO: use multierror like https://github.com/hashicorp/go-multierror
+	//hermes := mh.mid_hermes
+	//C.mid_hermes_destroy(&(hermes))
```

Comment:
shouldn't we destroy `mh.mid_hermes` here?

Reply:
mh.wrapper should call mid_hermes_destroy in `.Close()` because only he know how he initiated it
And one more problem that `mid_hermes_destroy` call *_destroy for 'stores' that assigned to `mh.mid_hermes` pointer which will be freed. And if `wrapper.Close()` will call mid_hermes_destroy too with own pointer of `mid_hermes` then will be dereferenced null pointer.

for example: 
mh.mid_hermes object has pointer with address 0x222 and points to midhermes struct at address 0xAAAA
wrapper has pointer with own address 0x333 and points to the same struct at address 0xAAA
we call mid_hermes_destroy and pass copy of pointer to 0x222. Then it's dereferenced and freed pointer to where it points (0xAAAA), and set 0x222 = NULL
then wrapper call mid_hermes_destroy and pass copy of pointer to 0x333, dereferenced and try to free 0xAAAA and took error

yeah, we can operate with **mid_hermes but more important that only wrapper know how it should be freed